### PR TITLE
perf: implement ViewHolder pattern in ContributorAdapter

### PR DESCRIPTION
**Pull request by HenriqueRocha**
* Add copyright message to the top of the file
* Implement ViewHolder pattern
* Use super.addAll if in API >= 10 because it doesn't call notifyDataSetChanged() for each add()
* Replace App.getInstance().getPicasso() with Picasso.with(context) since Picasso is already a singleton

### CODE REVIEW

**Comment by ViliusKraujutis** [comment]
lgtm :+1:

File: app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java [hunk 1]
```diff
@@ -1,60 +1,80 @@
...
     @Override
-    public View getView(int position, View view, ViewGroup parent) {
-        if(view == null)
-            view = mInflater.inflate(R.layout.contributor_item, null);
+    public View getView(int position, View convertView, ViewGroup parent) {
+        if (convertView == null) {
+            convertView = View.inflate(getContext(), R.layout.contributor_item, null);
+
+            ViewHolder viewHolder = new ViewHolder();
+            viewHolder.contributorName = (TextView) convertView.findViewById(R.id.contributorName);
```
Comment (ViliusKraujutis) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: Use ButterKnife to annotate `contributorName` field, and these could be removed.

File: app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java [hunk 2]
```diff
@@ -1,60 +1,80 @@
...
+        return convertView;
     }
 
     @Override
     public void addAll(Collection<? extends Contributor> contributors) {
-        for (Contributor c : contributors) {
-            add(c);
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.GINGERBREAD_MR1) {
+            super.addAll(contributors);
+        } else {
```
Comment (ViliusKraujutis) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: Need to add `setNotifyOnChange(false)` and then `setNotifyOnChange(true) after all items are added. And then notify it manually.

See implementation:
```
    /**
     * Adds the specified Collection at the end of the array.
     *
     * @param collection The Collection to add at the end of the array.
     */
    public void addAll(Collection<? extends T> collection) {
        synchronized (mLock) {
            if (mOriginalValues != null) {
                mOriginalValues.addAll(collection);
            } else {
                mObjects.addAll(collection);
            }
        }
        if (mNotifyOnChange) notifyDataSetChanged();
    }
```

and

```
    /**
     * Adds the specified object at the end of the array.
     *
     * @param object The object to add at the end of the array.
     */
    public void add(T object) {
        synchronized (mLock) {
            if (mOriginalValues != null) {
                mOriginalValues.add(object);
            } else {
                mObjects.add(object);
            }
        }
        if (mNotifyOnChange) notifyDataSetChanged();
    }
```

**Comment by HenriqueRocha** [comment]
:x:

**Comment by friedger** [comment]
? Don't :x: without a reason

**Comment by HenriqueRocha** [comment]
Sorry. Updated the comment to include reason. Everything should be okay now.

File: app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java [hunk 3]
```diff
@@ -1,60 +1,83 @@
...
-        name.setText(contributor.getLogin());
+        ViewHolder viewHolder = (ViewHolder) convertView.getTag();
+        viewHolder.contributorName.setText(contributor.getLogin());
+
         if (!TextUtils.isEmpty(contributor.getAvatarUrl())) {
-            final ImageView c = (ImageView) view.findViewById(R.id.contributorIcon);
-            App.getInstance().getPicasso()
-                    .load(contributor.getAvatarUrl())
-                    .into(c);
+            Picasso.with(getContext()).load(contributor.getAvatarUrl()).into(viewHolder.contributorAvatar);
```
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: Why are you not using the configure Picasso instance?

Comment (HenriqueRocha) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: I didn't notice that Picasso was being configured. I'll change that.

File: app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java [hunk 4]
```diff
@@ -15,46 +30,55 @@
...
     }
 
     @Override
     public void addAll(Collection<? extends Contributor> contributors) {
+        setNotifyOnChange(false);
         for (Contributor c : contributors) {
             add(c);
         }
+        setNotifyOnChange(true);
+        notifyDataSetChanged();
```
Comment (ViliusKraujutis) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: :+1: 
#perfmatters :)

Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: `notifyDataSetChanged` already calls `setNotifyOnChange(true);` internally. :)

**Comment by tasomaniac** [comment]
:+1:

File: app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java (hunk 4 above)
Comment (HenriqueRocha) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: Should I change the code or leave it like it is? Setting notifyOnChange to true twice is not that bad and I think it makes the code more symmetric and more readable for someone who doesn't know that notifyDataSetChange already does mNotifyOnChange = true;
Let me know if you want me to change it.

Comment (ViliusKraujutis) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: I suggest you can remove `setNotifyOnChange(true);` and add a comment on this line `notifyDataSetChanged();` saying `// also sets mNotifyOnChange = true;`.

**Comment by ViliusKraujutis** [comment]
:+1:

File: app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java (hunk 4 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: First of all thank you. :)
I think that it is not that important. Thank you for the comment @ViliusKraujutis

Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/adapter/ContributorAdapter.java]: For devs knowing that setNotifyOnChange is set implicitly it is not clear
why it is mentioned explicitly. I recommend to remove it #prefmatters


  [image: --]
Friedger Müffke
[image: http://]about.me/friedger
<http://about.me/friedger>


2015-02-05 17:57 GMT+01:00 Said Tahsin Dane <notifications@github.com>:

**Comment by tasomaniac** [comment]
OutOfMemoryException again on Travis again. How should be solve this?

**Comment by stefanhoth** [comment]
retest this please

**Comment by stefanhoth** [comment]
I'm trying to rerun the build now.

**Comment by HenriqueRocha** [comment]
Since the build in develop is now passing, do you want me to rebase my two PRs with develop and push again?

**Comment by stefanhoth** [comment]
@HenriqueRocha Yeah, let's try this first. Somehow the dex gets too big.

**Comment by HenriqueRocha** [comment]
@stefanhoth Strange, for some reason, after rebasing both PRs, the other one passed and this one failed in the dexing with OOM.

**Comment by stefanhoth** [comment]
I'm working on fixing the OOM hopefully.

**Pull request merged**